rapid7_insight_agent/icon_rapid7_insight_agent/util/graphql_api/api_connection.py

```diff
@@ -0,0 +1,234 @@
+from icon_rapid7_insight_agent.util.graphql_api.region_map import region_map
+from icon_rapid7_insight_agent.util.graphql_api.api_exception import APIException
+import icon_rapid7_insight_agent.util.graphql_api.agent_typer as agent_typer
+import requests
+
+class ApiConnection():
+    def __init__(self, api_key, org_key, region_string, logger):
+        self.api_key = api_key
+        self.org_key = org_key
+        self.logger = logger
+
+        self.endpoint = self._setup_endpoint(region_string)
+
+        self.session = requests.Session()
+        self.session.headers = self._get_headers()
+
+    def get_agent(self, agent_input):
+        agent_type = agent_typer.get_agent_type(agent_input)
+        agents = self._get_all_agents()
+        agent = self._find_agent_in_agents(agents, agent_input, agent_type)
+        return agent
+
+    def quarantine(self, advertisement_period, agent_id, quarantine_state):
+        """
+        Take a quarantine action on a given agent id
+
+        :param advertisement_period: int (Amount of time in seconds to try to take the quarantine action)
+        :param agent_id: string
+        :param quarantine_state: boolean (quarantine or unquarantine the agent)
+
+        :return: boolean
+        """
+        if quarantine_state:
+            quarantine_payload = {
+                "query": "mutation( $orgID:String! $agentID:String! $advPeriod:Long! ) { quarantineAssets( orgId:$orgID assetIds: [$agentID] input: {advertisementPeriod: $advPeriod} ) { results { assetId failed } } }",
+                "variables": {
+                    "orgID": self.org_key,
+                    "agentID": agent_id,
+                    "advPeriod": advertisement_period
+                }
+            }
+        else:
+            quarantine_payload = {
+                "query": "mutation( $orgID:String! $agentID:String!) { unquarantineAssets( orgId:$orgID assetIds: [$agentID] ) { results { assetId failed } } }",
+                "variables": {
+                    "orgID": self.org_key,
+                    "agentID": agent_id
+                }
+            }
+
+        action_verb = "quarantine" if quarantine_state else "unquarantine"
```

Comment:
Most of the `if quarantine_state`/etc logic throughout this function can be moved into the first if/else block. Alternatively, I could see this being broken out into two separate functions, one for quarantine and one for unquarantine. Those functions could be private and then called via a single function (`manage_quarantine_state`) or something like that.

Reply:
I split the functions into quarantine and unquarantine